python/drned_xmnr/op/filtering.py

```diff
@@ -356,184 +355,312 @@ def event_generator(consumer):
-    on `LineOutputEvent` instances.
+class LSMachine(object):
+    '''Simple state machine with a stack.
+
+    Input (events) are passed for handling to states.  A state handler
+    returns a 2-tuple (handled, states) or a 3-tuple (handled, states,
+    line).
+
+    handled: Bool - it True, the current event has been processed
+    and another one needs to be read (if True); otherwise it is used again.
+
+    states: [LS] - list of states to be used in place of the current state (an "expansion").
+
+    line: Str - output to be produced, if present; typically from `event.produce_line()`.
+
     '''
+
     def __init__(self, init_state):
         self.stack = [init_state]
 
     def handle(self, event):
-        while not event.complete and self.stack != []:
-            # print('handling', event, self.stack)
-            state = self.stack[-1]
-            result = state.handle(event)
-            if isinstance(result, tuple):
-                line, newstate = result
-                if newstate is TERMINATED:
-                    self.stack.pop()
-                    yield line
-                else:
-                    self.stack.append(newstate)
-                    yield line
-            elif isinstance(result, str):
-                yield result
-            elif isinstance(result, LogState):
-                self.stack.append(result)
-            elif result is TERMINATED:
-                self.stack.pop()
-
-
-class LogState(object):
+        handled = False
+        while not handled and self.stack:
+            # print('handling', event, [state.name for state in reversed(self.stack)])
+            state = self.stack.pop()
+            restuple = state.handle(event)
+            if len(restuple) == 2:
+                handled, expansion = restuple
+            else:
+                handled, expansion, line = restuple
+                yield line
+            self.stack.extend(reversed(expansion))
+
+
+class LS(object):
+    name = 'general logstate'
+
     def handle(self, event):
-        return event.produce_line()
+        return (False, [self], event.produce_line())
 
 
-class TransitionLogState(LogState):
+class TransitionTestS(LS):
+    'Tinitial state for "transition" action output filtering.'
+
+    name = 'transition-test'
+
     def __init__(self, level):
         self.level = level
-        self.initialized = False
 
     def handle(self, event):
         if self.level == 'overview':
-            event.mark_complete()
-            return None
-        if not self.initialized:
-            self.initialized = True
-            return (DrnedPrepare().produce_line(), DrnedLogState())
-        event.mark_complete()
-        return event.produce_line()
+            return (True, [self])
+        return (False, [TransitionS()], DrnedPrepare().produce_line())
+
+
+class TransitionS(LS):
+
+    '''
+    Transition -> (Commit)* Load Commit Compare (Rollback Commit Compare)*
```

Comment:
Should I understand this as a simple grammar? I guess yes because transition appear in similar rules below for explore and walk states... If yes I think it could make sense to maybe write non-terminals like 'Transition' in all upper case or something like that to make them stand out.

Reply:
I agree, this is not a simple grammar, but 1) that's the language we have to deal with (meaning DrNED output), 2) it is a at least a grammar now (contrary to how it looked before). I believe I write all non-terminals with a upper-case initial and terminals (events) all lower-case so that the non-terminals are pretty much state class names and terminals the event names.